[TOC]: #

# Table of Contents
- [Assess read accuracy from a PAF file](#assess-read-accuracy-from-a-paf-file)
- [Assess per-base accuracy](#assess-per-base-accuracy)
- [Iterative pilon polishing](#iterative-pilon-polishing)
- [Remove contaminant contigs](#remove-contaminant-contigs)


## Assess read accuracy from a PAF file
[`assess_accuracy_from_paf.py`][4]

```
Usage: read_accuracy_from_paf.py [OPTIONS]

  A script to produce data relevant to assessing the per-read accuracy for a
  PAF file. The output is a file with columns containing read identifier,
  read length, BLAST identity, and relative length of read compared to the
  reference.

Options:
  -h, --help             Show this message and exit.
  -i, --infile FILE      PAF file to assess.  [required]
  -o, --output FILENAME  The path to write the output file to. Use '-' for
                         stdout  [required]

  --delim TEXT           The column delimiter to use in the output file.
                         [default: ,]

  --primary-only         Only assess primary alignments.
  --min-cov INTEGER      Minimum read coverage required for a record to be
                         assessed. Read coverage is defined as the proportion
                         of the query sequence involved in the alignment. It
                         is the aligned length minus the read length.
                         [default: 0]

```

## Assess per-base accuracy
[`assess_per_base_accuracy.py`][1]

```
$ python assess_per_base_accuracy.py --help
Usage: assess_per_base_accuracy.py [OPTIONS]

  A script to assess the per-base quality of an assembly. There are two key
  metrics analysed here:

  1. Per-base consensus of reads mapped to the assembly. Consensus is
  defined as whether the percentage of reads matching the assembly at a site
  is greater than or equal to the value passed to --quorum

  2. Mapping quality summary statistics. i.e. mean, median, quantiles etc.

Options:
  -h, --help              Show this message and exit.
  --bam PATH              Bam file to assess. It is up to the user whether to
                          include secondary/supplementary alignments. These
                          can be removed with `samtools view -bh -F 3328`
                          [required]

  --pileup PATH           Pileup file generated by samtools mpileup. The
                          pileup MUST be generated with the --fasta-ref option
                          - with the assembly being provided as the fasta ref.
                          It is also recommended that the pileup is generated
                          using the mpileup flag -aa to report absolutely all
                          positions.  [required]

  --prefix TEXT           Path prefix to write output JSON and BED files to.
                          [required]

  --quorum INTEGER RANGE  Percentage of reads that must agree with the
                          assembly at each position.  [default: 95]

  --min-depth INTEGER     Minimum depth of reads required at a position. If
                          read depth at a position is below this value, the
                          position is considered in disagreement with the
                          assembly.  [default: 0]

  -v, --verbose           Turns on debug-level logging.
```

## Iterative pilon polishing
[`pilon_iterative.py`][2]

```
$ python pilon_iterative.py --help
Usage: pilon_iterative.py [OPTIONS]

  Iteratively run pilon to correct an assembly with Illumina reads. Stops
  after a specified number of iterations, or if no changes were made on the
  last iteration.

Options:
  -h, --help                      Show this message and exit.
  -1, --reads1 FILE               Name of input reads file 1.  [required]
  -2, --reads2 FILE               Name of input reads file 2.  [required]
  -o, --outdir DIRECTORY          Name of output/working directory.
                                  [required]

  -f, --final-fasta TEXT          Name of the final, polished fasta file.
                                  [default: final.fasta]

  -a, --assembly FILE             Name of input assembly fasta file.
                                  [required]

  -j, --pilon-jar FILE            Name of pilon JAR file  [required]
  -i, --max-iterations INTEGER RANGE
                                  Max number of iterations.  [default: 10]
  -t, --threads INTEGER RANGE     Number of threads to use with BWA mem and
                                  pilon.  [default: 1]

  -m, --pilon-memory TEXT         Maximum memory allocation pool for running
                                  Pilon. It is passed as `java -Xmx<value>`.
                                  Format: <int>[g|G|m|M|k|K]. See
                                  https://bit.ly/3c3KEot for more information.
                                  [default: 8G]

  --fix TEXT                      A comma-separated list of categories of
                                  issues to try to fix. Refer to the pilon
                                  documentation for valid values.  [default:
                                  all]

  --min-mapq INTEGER              Minimum alignment mapping quality for a read
                                  to count in pileups  [default: 10]

  --min-qual INTEGER              Minimum base quality to consider for pileups
                                  [default: 10]

  --force                         Ignore any previous checkpoints and polish
                                  from iteration 1.

  -l, --log-file TEXT             Name of the log file.  [default: log.txt]
```

## Remove contaminant contigs
[`remove_contamination.py`][3]

```
$ python remove_contamination.py --help
Usage: remove_contamination.py [OPTIONS]

Options:
  -t, --taxtree FILE         Taxon IDs NOT considered contamination. This file
                             can be either a taxon ID on each line, or a
                             taxonomic tree output by `taxonkit`.  [required]

  -S, --classification FILE  A centrifuge classification file for the input.
                             [required]

  -i, --input FILE           The fast{a,q} file to remove contamination from.
                             [required]

  -o, --output FILE          The path to write the decontaminated sequences
                             to.  [default: -; required]

  -v, --invert               Remove any record with a classification outside
                             the taxtree. By default, any record with a
                             classification within the taxtree is kept;
                             regardless of whether it has off-taxtree
                             classifications.

  -h, --help                 Show this message and exit.
  --verbose                  Turns on debug-level logging.
```



[1]: https://github.com/mbhall88/bioscripts/blob/master/python/assess_per_base_accuracy.py
[2]: https://github.com/mbhall88/bioscripts/blob/master/python/pilon_iterative.py
[3]: https://github.com/mbhall88/bioscripts/blob/master/python/remove_contamination.py
[4]: https://github.com/mbhall88/bioscripts/blob/master/python/assess_accuracy_from_paf.py
